### PR TITLE
Symfony3 Compatibility

### PR DESCRIPTION
**Pull request by hackzilla**
Currently in use with the (Password Generator App)[https://github.com/hackzilla/password-generator-app] which is Symfony v3

The only downside is this raises the minimum Symfony version 2.8

Thoughts?

/cc @Soullivaneuh

### CODE REVIEW

File: Controller/GeneratorController.php [hunk 1]
```diff
@@ -114,9 +113,12 @@ private function getMode(Request $request, $mode = null)
      */
     private function buildForm(PasswordGeneratorInterface $passwordGenerator, Options $options, $mode = '')
     {
-        return $this->createForm(new OptionType($passwordGenerator), $options, array(
-            'action' => $this->generateUrl('hackzilla_password_generator_show', array('mode' => $mode)),
+        return $this->createForm('Hackzilla\Bundle\PasswordGeneratorBundle\Form\Type\OptionType', $options, array(
```
Comment (Soullivaneuh) [inline Controller/GeneratorController.php]: If you bump minimum PHP requirement to 5.5, you could use `OptionType::class` that is nicer to read.

File: Form/Type/OptionType.php [hunk 2]
```diff
@@ -92,22 +100,11 @@ public function configureOptions(OptionsResolver $resolver)
...
-        $this->configureOptions($resolver);
-    }
-
-    /**
-     * @return string
-     */
-    public function getName()
+    public function getBlockPrefix()
     {
         return '';
```
Comment (Soullivaneuh) [inline Form/Type/OptionType.php]: Why an empty name/prefix?

File: README.md [hunk 3]
```diff
@@ -8,6 +8,7 @@ Simple multilingual bundle to add to any project. Available languages (Pull Requ
 * English
 * French
 
+
```
Comment (Soullivaneuh) [inline README.md]: Not needed new line IMO.

**Comment by Soullivaneuh** [comment]
Globally :+1: for this.

Concerning version support, keeping SF 2.3 support without deprecation is not so hard and this version is still maintained for the moment (until 2017 AFAIK).

I can understand SF 2.3 dropping, but I would suggest you to keep 2.7 support because this is a LTS version too.

File: Form/Type/OptionType.php (hunk 2 above)
Comment (hackzilla) [inline Form/Type/OptionType.php]: The prefix was being added to the url parameters.

This was the only way I found to keep the parameters simple

Comment (hackzilla) [inline Form/Type/OptionType.php]: http://password.dev/en/computer?quantity=5&uppercase=1&lowercase=1&numbers=1&avoid_similar=1&length=12

vs

http://password.dev/en/computer?option%5Bquantity%5D=5&option%5Buppercase%5D=1&option%5Blowercase%5D=1&option%5Bnumbers%5D=1&option%5Bavoid_similar%5D=1&option%5Blength%5D=12

File: Controller/GeneratorController.php (hunk 1 above)
Comment (hackzilla) [inline Controller/GeneratorController.php]: :+1:

**Comment by hackzilla** [comment]
I've reworked the PR, and allowed Symfony 2.7, but didn't get chance to check if it worked.

File: Form/Type/OptionType.php (hunk 2 above)
Comment (Soullivaneuh) [inline Form/Type/OptionType.php]: Hmm I see and I didn't know if we can do it in a better way.

Ok then. :+1:

File: Form/Type/OptionType.php [hunk 4]
```diff
@@ -50,37 +46,52 @@ public function buildForm(FormBuilderInterface $builder, array $options)
...
     private function addStringType(FormBuilderInterface $builder, $key, OptionInterface $option)
     {
         $builder->add(
-            $builder->create(strtolower($key), 'text', array(
-                'data' => $option->getValue(),
-                'label' => 'OPTION_'.$key,
-                'required' => false,
-            ))
+            $builder->create(strtolower($key), TextType::class, [
+                    'data'     => $option->getValue(),
```
Comment (Soullivaneuh) [inline Form/Type/OptionType.php]: Bad indent.

File: Form/Type/OptionType.php [hunk 5]
```diff
@@ -50,37 +46,52 @@ public function buildForm(FormBuilderInterface $builder, array $options)
...
     {
         $builder->add(
-            $builder->create(strtolower($key), 'checkbox', array(
-                'value' => 1,
-                'data' => $option->getValue(),
-                'label' => 'OPTION_'.$key,
-                'required' => false,
-            ))
+            $builder->create(strtolower($key), CheckboxType::class, [
+                    'value'    => 1,
```
Comment (Soullivaneuh) [inline Form/Type/OptionType.php]: Bad indent.

File: Form/Type/OptionType.php [hunk 6]
```diff
@@ -50,37 +46,52 @@ public function buildForm(FormBuilderInterface $builder, array $options)
...
     private function addIntegerType(FormBuilderInterface $builder, $key, OptionInterface $option)
     {
         $builder->add(
-            $builder->create(strtolower($key), 'integer', array(
-                'data' => $option->getValue(),
-                'label' => 'OPTION_'.$key,
-                'required' => false,
-            ))
+            $builder->create(strtolower($key), IntegerType::class, [
+                    'data'     => $option->getValue(),
```
Comment (Soullivaneuh) [inline Form/Type/OptionType.php]: Bad indent.

File: Form/Type/OptionType.php (hunk 2 above)
Comment (Soullivaneuh) [inline Form/Type/OptionType.php]: If you support SF 2.7, you have to keep the `getName` method:

```php

    /**
     * {@inheritdoc}
     * 
     * SF < 2.8 BC
     */
    public function getName()
    {
        return $this->getBlockPrefix();
    }

    /**
     * {@inheritdoc}
     */
    public function getBlockPrefix()
    {
        return '';
    }
```

See: https://github.com/symfony/symfony/blob/2.8/UPGRADE-2.8.md#form

File: Form/Type/OptionType.php [hunk 7]
```diff
@@ -89,25 +100,16 @@ private function addIntegerType(FormBuilderInterface $builder, $key, OptionInter
...
-
-
-    /**
-     * {@inheritdoc}
-     */
-    public function setDefaultOptions(OptionsResolverInterface $resolver)
-    {
-        $this->configureOptions($resolver);
+        $resolver->setDefaults(
+            array(
```
Comment (Soullivaneuh) [inline Form/Type/OptionType.php]: Short array syntax could be used here.

BTW, you should change [`.php_cs`](https://github.com/hackzilla/password-generator-bundle/blob/e740565f6e769bf095b6a06f9c94a280fd651d16/.php_cs#L13) configuration.

File: Form/Type/OptionType.php [hunk 8]
```diff
@@ -6,34 +6,30 @@
...
     {
         $builder
-            ->add('quantity', 'integer', array(
-                'label' => 'OPTION_HOW_MANY_PASSWORDS',
-            ));
+            ->add('quantity', IntegerType::class, [
+                    'label' => 'OPTION_HOW_MANY_PASSWORDS',
+            ]);
+
+        if (!is_a($options['generator'], 'Hackzilla\PasswordGenerator\Generator\PasswordGeneratorInterface')) {
```
Comment (Soullivaneuh) [inline Form/Type/OptionType.php]: `PasswordGeneratorInterface::class`

Comment (Soullivaneuh) [inline Form/Type/OptionType.php]: Or:

```php
$options['generator'] instanceof PasswordGeneratorInterface
```

File: Form/Type/OptionType.php [hunk 9]
```diff
@@ -89,25 +100,16 @@ private function addIntegerType(FormBuilderInterface $builder, $key, OptionInter
...
-
-    /**
-     * {@inheritdoc}
-     */
-    public function setDefaultOptions(OptionsResolverInterface $resolver)
-    {
-        $this->configureOptions($resolver);
+        $resolver->setDefaults(
+            array(
+                'data_class'      => 'Hackzilla\Bundle\PasswordGeneratorBundle\Entity\Options',
```
Comment (Soullivaneuh) [inline Form/Type/OptionType.php]: `Options::class`

**Comment by hackzilla** [comment]
Thanks for all the suggestions.

**Comment by hackzilla** [comment]
The original reason I dropped Symfony 2.7 was because the current errors: ```Could not load type "Symfony\Component\Form\Extension\Core\Type\IntegerType"```

I'm currently looking at a way to see if I can get this to work, otherwise I'll have to drop 2.7 support in 2.x

**Comment by Soullivaneuh** [comment]
Where is this error and how to reproduce it?

File: composer.json [hunk 10]
```diff
@@ -13,16 +13,21 @@
     "require": {
         "php": ">=5.3.2",
```
Comment (Soullivaneuh) [inline composer.json]: Must be bumped to `>=5.5` if you use short array syntax.

BTW the better way is `^5.5 || ^7.0`.

Comment (hackzilla) [inline composer.json]: :+1:

**Comment by hackzilla** [comment]
The travis tests produce it. https://travis-ci.org/hackzilla/password-generator-bundle/jobs/100440203

In OptionType.php, when you specify the FQCN, Symfony 2.7 throws this error.

```
        $builder
            ->add(
                'quantity',
                IntegerType::class,
                [
                    'label' => 'OPTION_HOW_MANY_PASSWORDS',
                ]
            );
```

Previously working code for Symfony 2.x

```
        $builder
            ->add('quantity', 'integer', array(
                'label' => 'OPTION_HOW_MANY_PASSWORDS',
            ));
```

**Comment by Soullivaneuh** [comment]
This is normal. FQDN argument is available since SF 2.8 only.

You have to conditional loading, with a `method_exists(AbstractType::type, 'getBlockPrefix')` for example (this method was introduced on SF 2.8 too.).

See: https://github.com/symfony/symfony/blob/2.8/UPGRADE-2.8.md#form

**Comment by hackzilla** [comment]
That's better than the path I was headed down using Symfony::VERSION

**Comment by Soullivaneuh** [comment]
Yes, because Symfony version constant **is only** for the kernel component, not the full stack.

**Comment by hackzilla** [comment]
I'll give this another look through tonight, but I think we're there for v2

File: Form/Type/OptionType.php [hunk 11]
```diff
@@ -2,38 +2,39 @@
...
     public function buildForm(FormBuilderInterface $builder, array $options)
     {
         $builder
-            ->add('quantity', 'integer', array(
-                'label' => 'OPTION_HOW_MANY_PASSWORDS',
-            ));
+            ->add(
+                'quantity',
+                method_exists(AbstractType::class, 'getBlockPrefix') ? IntegerType::class : 'integer',
+                [
```
Comment (Soullivaneuh) [inline Form/Type/OptionType.php]:
```php
method_exists(AbstractType::class, 'getBlockPrefix') ? IntegerType::class : 'integer', [
    // stuff,
]
```

More convenient, it is not?

File: Form/Type/OptionType.php [hunk 12]
```diff
@@ -89,25 +117,21 @@ private function addIntegerType(FormBuilderInterface $builder, $key, OptionInter
      */
     public function configureOptions(OptionsResolver $resolver)
     {
-        $resolver->setDefaults(array(
-            'data_class' => 'Hackzilla\Bundle\PasswordGeneratorBundle\Entity\Options',
-            'csrf_protection' => false,
-        ));
+        $resolver->setDefaults(
+            [
```
Comment (Soullivaneuh) [inline Form/Type/OptionType.php]: Same here, you can do nicer with starting array on function begin.

File: README.md [hunk 13]
```diff
@@ -27,7 +36,7 @@ Add HackzillaPasswordGeneratorBundle in your composer.json:
 ```yaml
 {
     "require": {
-        "hackzilla/password-generator-bundle": "~1.0"
+        "hackzilla/password-generator-bundle": "~2.0"
```
Comment (Soullivaneuh) [inline README.md]: The new convention is `^2.0`. `~` is not *really* semver compliant.

File: composer.json [hunk 14]
```diff
@@ -11,18 +11,23 @@
         }
     ],
     "require": {
-        "php": ">=5.3.2",
+        "php": "^5.5 || ^7.0",
         "hackzilla/password-generator": "~1.0",
-        "symfony/framework-bundle": "~2.3|~3.0",
-        "symfony/form": "~2.3|~3.0"
+        "symfony/framework-bundle": "~2.7|~3.0",
```
Comment (Soullivaneuh) [inline composer.json]: The new recommended composer syntax is `^2.7 || ^3.0`.

You should take profit to update all.

**Comment by Soullivaneuh** [comment]
Some last things:

* Some array syntaxes could be improved for nicer reading IMO.
* You still have long array syntaxes. I would suggest you to run php-cs-fixer. BTW, I'll propose you StyleCI integration for PR.

**Comment by hackzilla** [comment]
Cheers, I'll work through the StyleCI tonight.

**Comment by Soullivaneuh** [comment]
Oh I see you added StyleCI bot.

You have to configure fixer with `.styleci.yml` file.

Please take a look to my project to avoid duplicate configuration: https://github.com/Soullivaneuh/php-cs-fixer-styleci-bridge

**Pull request merged**